packages/overture-schema-system/src/overture/schema/system/constraint/string.py

```diff
@@ -0,0 +1,376 @@
+import re
+from typing import Any
+
+from pydantic import (
+    GetCoreSchemaHandler,
+    GetJsonSchemaHandler,
+    ValidationError,
+    ValidationInfo,
+)
+from pydantic_core import InitErrorDetails, core_schema
+
+from overture.schema.system.constraint import (
+    Constraint,
+)
+
+
+class StringConstraint(Constraint):
+    """Base class for string-based constraints."""
+
+    def __get_pydantic_core_schema__(
+        self, source: type[Any], handler: GetCoreSchemaHandler
+    ) -> core_schema.CoreSchema:
+        python_schema = handler(str)
+
+        def validate_string(value: str, info: ValidationInfo) -> str:
+            self.validate(value, info)
+            return value
+
+        return core_schema.with_info_after_validator_function(
+            validate_string, python_schema
+        )
+
+
+class PatternConstraint(StringConstraint):
+    """Generic pattern-based string constraint."""
+
+    def __init__(self, pattern: str, error_message: str, flags: int = 0):
+        self.pattern = re.compile(pattern, flags)
+        self.pattern_str = pattern
```

Comment:
This looks like it can go away, esp. since it's not actually a string representation.

Reply:
`self.pattern_str` is unnecessary; the lone reference to it (see the next comment) can be changed.